internal/controllers/gateway/gateway_utils.go

```diff
@@ -300,12 +292,42 @@ func getListenerStatus(
 			}
 		}
 
+		// gets all the HTTPRoutes and update the listeners' attachedRoutes field.
+		// At this point we take into account HTTPRoutes only, as they are the
+		// only routes in beta.
+		httpRouteList := gatewayv1beta1.HTTPRouteList{}
+		if err := client.List(ctx, &httpRouteList); err != nil {
+			return nil, err
+		}
+		var attachedRoutes int32
+		for _, route := range httpRouteList.Items {
+			route := route
+			if !isRouteAcceptedByGateway(route.Namespace, route.Status.Parents, *gateway) {
+				continue
+			}
+			for _, parentRef := range route.Spec.ParentRefs {
+				accepted, err := isRouteAcceptedByListener(
+					ctx,
+					client,
+					&route,
+					listener,
+					*gateway,
+					parentRef,
+				)
+				if err != nil {
+					return nil, err
+				}
+				if accepted {
+					attachedRoutes++
+				}
+			}
+		}
+
```

Comment:
The `getListenerStatus` function is becoming a bit huge, what about encapsulating this logic into it's own function?

Maybe:

```suggestion
		attachedRoutes, err := getAttachedRoutesForGateway(ctx, gateway)
		if err != nil {
			return nil, err
		}
```

Reply:
Yeah, it makes sense

Reply:
31dacf82adffcaa4434eeb2f28f71bad0dc65e90